libyul/ASTNodeRegistry.cpp

```diff
@@ -0,0 +1,150 @@
+}
+
+size_t ASTNodeRegistry::idToLabelIndex(NodeId const _id) const
+{
+	yulAssert(_id < m_idToLabelMapping.size());
+	return m_idToLabelMapping[_id];
+}
+
+std::string_view ASTNodeRegistry::operator()(NodeId const _id) const
+{
+	auto const labelIndex = idToLabelIndex(_id);
+	if (labelIndex == ghostId())
+		return lookupGhost(_id);
+	return m_labels[labelIndex];
+}
+
+std::optional<ASTNodeRegistry::NodeId> ASTNodeRegistry::findIdForLabel(std::string_view const _label) const {
+	if (_label.empty())
+		return emptyId();
+	if (_label == ghostPlaceholder)
+		return NodeId{1};
+	for (NodeId id = 2; id <= maximumId(); ++id)
+		if ((*this)(id) == _label)
+			return id;
+	return std::nullopt;
+}
+
+std::string_view ASTNodeRegistry::lookupGhost(NodeId const _id) const
+{
+	yulAssert(idToLabelIndex(_id) == ghostId());
+	auto const [it, _] = m_ghostLabelCache.try_emplace(_id, fmt::format("GHOST[{}]", _id));
+	return it->second;
+}
+
+ASTNodeRegistryBuilder::LabelToIdMapping::LabelToIdMapping():
+	m_mapping{{"", 0}, {ASTNodeRegistry::ghostPlaceholder, 1}}
+{}
+
+std::tuple<ASTNodeRegistry::NodeId, bool> ASTNodeRegistryBuilder::LabelToIdMapping::tryInsert(
+	std::string_view const _label,
+	ASTNodeRegistry::NodeId const _id
+)
+{
+	yulAssert(_label != ASTNodeRegistry::ghostPlaceholder);
+	auto const [it, emplaced] = m_mapping.try_emplace(std::string{_label}, _id);
+	return std::make_tuple(it->second, emplaced);
+}
+
+ASTNodeRegistryBuilder::ASTNodeRegistryBuilder():
+	m_nextId(2)
+{}
+
+ASTNodeRegistryBuilder::ASTNodeRegistryBuilder(ASTNodeRegistry const& _existingRegistry)
+{
+	for (size_t i = 2; i <= _existingRegistry.maximumId(); ++i)
+	{
+		auto const existingLabel = _existingRegistry(i);
+		if (!existingLabel.empty())
+		{
+			auto const [_, inserted] = m_mapping.tryInsert(_existingRegistry(i), i);
+			yulAssert(inserted);
+		}
+	}
+	m_nextId = _existingRegistry.maximumId() + 1;
+}
+
+ASTNodeRegistry::NodeId ASTNodeRegistryBuilder::define(std::string_view const _label)
+{
+	auto const [id, inserted] = m_mapping.tryInsert(_label, m_nextId);
+	if (inserted)
+		m_nextId++;
+	return id;
+}
+
+ASTNodeRegistry ASTNodeRegistryBuilder::build() const
+{
+	auto const& mapping = m_mapping.data();
+	yulAssert(mapping.contains(""));
+	yulAssert(mapping.at("") == 0);
+	yulAssert(mapping.contains(ASTNodeRegistry::ghostPlaceholder));
+	yulAssert(mapping.at(ASTNodeRegistry::ghostPlaceholder) == 1);
+
+	std::vector<std::string> labels{"", ASTNodeRegistry::ghostPlaceholder};
+	labels.reserve(mapping.size());
+	std::vector<size_t> idToLabelMapping(ranges::max(mapping | ranges::views::values) + 1, 0);
+	yulAssert(idToLabelMapping.size() >= 2, "Mapping at least contains empty and ghost entry");
+	idToLabelMapping[1] = 1;
+	for (auto const& [label, id]: mapping)
+	{
+		// skip empty and ghost
+		if (id < 2)
+			continue;
+
+		labels.emplace_back(label);
+		idToLabelMapping[id] = labels.size() - 1;
+	}
+	return ASTNodeRegistry{labels, idToLabelMapping};
```

Comment:
I see that `ASTNodeRegistry` currently takes these by reference. Would be bit more efficient if it took them by value. Then we could just move already constructed containers:

```suggestion
	return ASTNodeRegistry{std::move(labels), std::move(idToLabelMapping)};
```

Reply:
Fixed!